plugins/lines/prism-lines.js

```diff
@@ -0,0 +1,130 @@
+(function () {
+
+	if (typeof self === 'undefined' || !self.Prism || !self.document) {
+		return;
+	}
+
+	/**
+	 * Returns style declarations for the element
+	 * @param {Element} element
+	 */
+	Prism.hooks.add('before-insert', function (env) {
+
+		// works only for <code> wrapped inside <pre> (not inline)
+		var pre = env.element.parentNode;
+		if (!pre || !/pre/i.test(pre.nodeName)) {
+			return;
+		}
+
+		// we are not welcome here :(
+		if (pre.className.includes('nolines')) {
+			return;
+		}
+
+		// support for line-numbers behaviour
+		var dataStart = 1;
+		if (pre.hasAttribute('data-start')) {
+			dataStart = parseInt(pre.getAttribute('data-start'), 10);
+			if (isNaN(dataStart)) {
+				dataStart = 1;
+			}
+			pre.style.setProperty('--counter-start', dataStart - 1);
+                }
+
+		function createLine(lineNumber, prefix) {
+			var line = document.createElement('div');
+			var anchor = document.createElement('a');
+			if (prefix) {
+				var target = prefix + '.' + lineNumber;
+				line.id = target;
+				anchor.href = '#' + target;
+			}
+			line.setAttribute('class', 'line');
+			line.appendChild(anchor);
+			return line;
+		}
+
+		function splitNode(node) {
+			if (node.hasChildNodes()) {
+				// so we've got a kid! :)
+				var p1 = document.createElement(node.nodeName);
+				var p2 = document.createElement(node.nodeName);
+				if (node.hasAttributes()) {
+					for (i = 0; i < node.attributes.length ; i++) {
+						var key = node.attributes.item(i).name;
+						var value = node.attributes.item(i).value;
+						p1.setAttribute(key, value);
+						p2.setAttribute(key, value);
+					}
+				}
+				var kid = node.firstChild;
+
+				// skip all non-interesting kids
+				while (kid && !/\n/.test(kid.textContent)) {
+					p1.appendChild(kid.cloneNode(true));
+					kid = kid.nextSibling;
+				}
+
+				if (kid && /\n/.test(kid.textContent)) {
+					// found a special one :)
+					var pair = splitNode(kid);
+					p1.appendChild(pair[0]);
+					if (pair[1])
+						p2.appendChild(pair[1]);
+				}
+
+				return [ p1, p2 ];
+			}
+
+			// we are at the bottom of the hierarchy
+			switch (node.nodeType) {
+				case Node.TEXT_NODE:
+					var newLine = node.textContent.indexOf('\n');
+					if (newLine == -1 || newLine == node.textContent.length - 1)
+						return [ node, null ];
+					return [
+						document.createTextNode(node.textContent.slice(0, newLine + 1)),
+						document.createTextNode(node.textContent.slice(newLine + 1))
+					];
+					break;
+				default:
+					console.warn('Lines plugin: encountered unexpected node type ', node.nodeType, ', please report this as bug.');
+					return [ node, null ]; // no splitting, safe
+			}
+		}
+
+		var parser = new DOMParser();
+		var highlightedCode = parser.parseFromString(env.highlightedCode, 'text/html');
+		var wrapped = new DocumentFragment();
```

Comment:
One thing to consider that not all options are equivalent.  The reason I chose Fragment that even if I recall correctly it is not susceptible to XSS attacks, while many other options are.

Also, keep in mind that if you just use `document.createElement()` you will need to handle many edge cases since not everything will be maintained in that buffer.  An example that comes to mind, although is artificial for Prism, is `<td>text</td>` in the input -- with `document.createElement()` storage tags will be stripped off and lost.  I wanted a storage that would represent whatever I am getting from `env.highlightCode`.

Reply:
However, I agree, that if we care about IE11, then `documentFragment()` is no go, since it is not supported by any IE.

Reply:
XSS is not a problem the DOMParser [won't execute `script` tags](https://stackoverflow.com/questions/28112807/why-script-elements-created-through-domparser-do-not-execute).

The edge cases are interesting. Is there a complete list of stripped tags? It can't be too many but it would be good to know.